File: backend/services/stigmer-server/pkg/controllers/workflowexecution/stream_broker.go
package workflowexecution

import (
	"sync"

	"github.com/rs/zerolog/log"
	workflowexecutionv1 "github.com/stigmer/stigmer/internal/gen/ai/stigmer/agentic/workflowexecution/v1"
)

// StreamBroker manages in-memory Go channels for real-time execution updates
//
// This implements the "Stream Broker" responsibility from ADR 011:
// > Stream Broker: Manages in-memory Go Channels to broadcast real-time updates to CLI watchers.
//
// Architecture:
// - When UpdateStatus is called, the broker broadcasts to all active subscribers
// - When Subscribe is called, a new channel is created and registered
// - When a subscriber disconnects, the channel is cleaned up
//
// This eliminates the need for polling and provides near-instant updates as per ADR.
type StreamBroker struct {
	mu          sync.RWMutex
	subscribers map[string][]chan *workflowexecutionv1.WorkflowExecution
}

// NewStreamBroker creates a new StreamBroker instance
func NewStreamBroker() *StreamBroker {
	return &StreamBroker{
		subscribers: make(map[string][]chan *workflowexecutionv1.WorkflowExecution),
	}
}

// Subscribe creates a new channel for the given execution ID and registers it
//
// The returned channel will receive all future updates for this execution.
// The caller MUST call Unsubscribe when done to prevent channel leaks.
//
// Returns:
// - A channel that receives WorkflowExecution updates
func (b *StreamBroker) Subscribe(executionID string) chan *workflowexecutionv1.WorkflowExecution {
	b.mu.Lock()
	defer b.mu.Unlock()

	// Create buffered channel to prevent blocking broadcasts
	// Buffer size of 100 allows bursts of updates without blocking
	ch := make(chan *workflowexecutionv1.WorkflowExecution, 100)

	// Register subscriber
	b.subscribers[executionID] = append(b.subscribers[executionID], ch)

	log.Debug().
		Str("execution_id", executionID).
		Int("total_subscribers", len(b.subscribers[executionID])).
		Msg("New subscriber registered")

	return ch
}

// Unsubscribe removes a channel from the subscriber list and closes it
//
// This should be called when:
// - The client disconnects
// - The execution reaches a terminal state
// - The subscription is cancelled
func (b *StreamBroker) Unsubscribe(executionID string, ch chan *workflowexecutionv1.WorkflowExecution) {
	b.mu.Lock()
	defer b.mu.Unlock()

	subscribers, exists := b.subscribers[executionID]
	if !exists {
		return
	}

	// Find and remove the channel
	for i, subscriber := range subscribers {
		if subscriber == ch {
			// Remove from slice
			b.subscribers[executionID] = append(subscribers[:i], subscribers[i+1:]...)

			// Close the channel
			close(ch)

			log.Debug().
				Str("execution_id", executionID).
				Int("remaining_subscribers", len(b.subscribers[executionID])).
				Msg("Subscriber unregistered")

			// Clean up empty entries
			if len(b.subscribers[executionID]) == 0 {
				delete(b.subscribers, executionID)
			}

			break
		}
	}
}

// Broadcast sends an execution update to all active subscribers
//
// This is called after UpdateStatus persists to the database.
// It implements the "Daemon (Streaming): Pushes message to active Go Channels" step from ADR 011.
//
// The broadcast is non-blocking - if a channel's buffer is full, the update is dropped
// for that subscriber (they'll get the next one).
func (b *StreamBroker) Broadcast(execution *workflowexecutionv1.WorkflowExecution) {
	if execution == nil || execution.Metadata == nil {
		return
	}

	executionID := execution.Metadata.Id

	b.mu.RLock()
	defer b.mu.RUnlock()

	subscribers, exists := b.subscribers[executionID]
	if !exists || len(subscribers) == 0 {
		// No subscribers, nothing to do
		return
	}

	log.Debug().
		Str("execution_id", executionID).
		Str("phase", execution.Status.GetPhase().String()).
		Int("subscribers", len(subscribers)).
		Msg("Broadcasting execution update")

	// Broadcast to all subscribers
	for _, ch := range subscribers {
		select {
		case ch <- execution:
			// Successfully sent
		default:
			// Channel buffer full, drop this update
			// Subscriber will get the next update
			log.Warn().
				Str("execution_id", executionID).
				Msg("Subscriber channel full, dropping update")
		}
	}
}

// GetSubscriberCount returns the number of active subscribers for an execution
//
// This is useful for debugging and monitoring.
func (b *StreamBroker) GetSubscriberCount(executionID string) int {
	b.mu.RLock()
	defer b.mu.RUnlock()

	return len(b.subscribers[executionID])
}

File: backend/services/stigmer-server-ts/src/domain/agentexecution/stream-broker.ts
/**
 * StreamBroker — ports controller/stream_broker.go: the in-memory
 * broadcast fabric for real-time execution updates (ADR 011's "Stream
 * Broker" responsibility; the OSS stand-in for cloud's Redis streams).
 *
 * Go's shape is a map of executionID → buffered channels (capacity 100)
 * under an RWMutex: Broadcast is non-blocking — a full buffer drops the
 * frame for that subscriber (they catch up on the next one) — and
 * Unsubscribe closes the channel. The TS port keeps the exact delivery
 * semantics on the queue-plus-notify idiom the transport already proved
 * for streams (transport/health.ts watch): each subscription owns a
 * bounded FIFO queue the subscribe generator drains, and a notify hook
 * wakes the drain loop. Node's single-threaded event loop replaces the
 * mutex; the bounded queue replaces the channel buffer.
 *
 * ONE instance serves both routers (serving + in-process): #18's Temporal
 * activities update status through the in-process client, and those
 * broadcasts must reach externally-connected subscribers — the same
 * reason Go exposes GetStreamBroker to its activities. The composition
 * root owns the instance.
 */
import type { AgentExecution } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/api_pb";

import type { Logger } from "../../boot/logger.js";

/**
 * Go's channel buffer size: bursts up to this many undelivered frames per
 * subscriber are absorbed; beyond it, broadcasts drop for that subscriber
 * (stream_broker.go's `make(chan ..., 100)` + non-blocking send).
 */
export const SUBSCRIBER_BUFFER_CAPACITY = 100;

/**
 * One subscriber's delivery state. The subscribe stream loop drains
 * `queue` and parks on `notify`; the broker pushes and wakes. `closed`
 * flips on unsubscribe (Go's close(ch)) so a parked or late consumer
 * observes the end instead of waiting forever.
 */
export interface BrokerSubscription {
  readonly queue: AgentExecution[];
  notify: (() => void) | undefined;
  closed: boolean;
}

export class StreamBroker {
  private readonly subscribers = new Map<string, Set<BrokerSubscription>>();

  constructor(private readonly logger: Logger) {}

  /**
   * Registers a new subscriber for the execution. The caller MUST
   * unsubscribe when done (the subscribe generator's `finally`) — Go's
   * "caller MUST call Unsubscribe" contract, enforced there by defer.
   */
  subscribe(executionId: string): BrokerSubscription {
    const subscription: BrokerSubscription = {
      queue: [],
      notify: undefined,
      closed: false,
    };
    const set = this.subscribers.get(executionId) ?? new Set();
    set.add(subscription);
    this.subscribers.set(executionId, set);
    this.logger.debug("New subscriber registered", {
      executionId,
      totalSubscribers: set.size,
    });
    return subscription;
  }

  /** Removes and closes a subscription; idempotent (Go Unsubscribe). */
  unsubscribe(executionId: string, subscription: BrokerSubscription): void {
    const set = this.subscribers.get(executionId);
    if (set === undefined || !set.has(subscription)) {
      return;
    }
    set.delete(subscription);
    subscription.closed = true;
    subscription.notify?.();
    subscription.notify = undefined;
    this.logger.debug("Subscriber unregistered", {
      executionId,
      remainingSubscribers: set.size,
    });
    if (set.size === 0) {
      this.subscribers.delete(executionId);
    }
  }

  /**
   * Delivers an execution update to every active subscriber of its id.
   * Non-blocking by construction: a subscriber at buffer capacity drops
   * THIS frame (it catches up on the next broadcast) — Go's select/default
   * arm, warn included.
   */
  broadcast(execution: AgentExecution): void {
    const executionId = execution.metadata?.id ?? "";
    if (executionId === "") {
      return;
    }
    const set = this.subscribers.get(executionId);
    if (set === undefined || set.size === 0) {
      return;
    }
    this.logger.debug("Broadcasting execution update", {
      executionId,
      subscribers: set.size,
    });
    for (const subscription of set) {
      if (subscription.queue.length >= SUBSCRIBER_BUFFER_CAPACITY) {
        this.logger.warn("Subscriber channel full, dropping update", {
          executionId,
        });
        continue;
      }
      subscription.queue.push(execution);
      subscription.notify?.();
      subscription.notify = undefined;
    }
  }

  /** Active subscriber count for an execution (Go GetSubscriberCount). */
  getSubscriberCount(executionId: string): number {
    return this.subscribers.get(executionId)?.size ?? 0;
  }
}
